File: ARGBookUI/ARGBookDocumentContentView/Layouts/ARGBookDocumentHorizontalLayout/SettingsController/ARGFlowableLayoutSettingsController.m
//
//  BookContentViewSettings.m
//  Auri
//
//  Created by Sergei Polshcha on 16/04/16.
//

#import "ARGFlowableLayoutSettingsController.h"
#import <ARGBookUI/ARGBookUI-Swift.h>

@interface ARGFlowableLayoutSettingsController ()

@property (nonatomic, strong, readonly) NSString *pageSettingsString;

@end

@implementation ARGFlowableLayoutSettingsController

- (void)setSettings:(id<ARGBookReadingSettings>)settings completion:(dispatch_block_t)completion {
    __weak typeof (self) wself = self;
    
    [super setSettings:settings completion:^{
        dispatch_group_t settingsGroup = dispatch_group_create();
        
        dispatch_group_notify(settingsGroup, dispatch_get_main_queue(), ^{
            if (completion) {
                completion();
            }
        });
        
        dispatch_group_enter(settingsGroup);
        [wself setRelativePageMargins:UIOffsetMake(settings.horizontalMargin, settings.verticalMargin) completion:^{
            dispatch_group_leave(settingsGroup);
        }];
    }];
}

- (void)setRelativePageMargins:(UIOffset)pageMargins completion:(dispatch_block_t)completion {
    _relativePageMargins = pageMargins;
    
    UIOffset absolutePageMargins = UIOffsetMake(floor(self.pageSize.width / 100 * pageMargins.horizontal), floor(self.pageSize.height / 100 * pageMargins.vertical));
    
    [self setAbsolutePageMargins:absolutePageMargins completion:completion];
}

- (void)setAbsolutePageMargins:(UIOffset)pageMargins completion:(dispatch_block_t)completion {
    _absolutePageMargins = pageMargins;
    
    NSString *pageSizeScript = [NSString stringWithFormat:@"setPageSettings(%f, %f, %f, %f, %f, %f)",
                                floor(self.pageSize.width - pageMargins.horizontal * 2),
                                floor(self.pageSize.height - pageMargins.vertical * 2),
                                floor(pageMargins.vertical),
                                floor(pageMargins.horizontal),
                                floor(pageMargins.vertical),
                                floor(pageMargins.horizontal)];
    
    if (![pageSizeScript isEqualToString:_pageSettingsString]) {
        _pageSettingsString = pageSizeScript;
        
        [self.webView evaluateJavaScript:pageSizeScript completionHandler:^(id _Nullable result, NSError * _Nullable error) {
            if (completion) {
                completion();
            }
        }];
    } else {
        if (completion) {
            completion();
        }
    }
}

@end

File: ARGBookUI/ARGBookDocumentContentView/Layouts/ARGBookDocumentHorizontalLayout/SettingsController/ARGFlowableLayoutSettingsController.swift
//
//  File.swift
//  ARGBookUI
//
//  Created by Sergei Polshcha on 03.12.2020.
//

import Foundation
import UIKit

class ARGFlowableLayoutSettingsController : ARGBookReadingSettingsController {
    
    private(set) var pageSettingsString: String?
    
    override func setSettings(_ settings: ARGBookReadingSettings, pageSize: CGSize, completionHandler: (() -> Void)? = nil) {
        super.setSettings(settings, pageSize: pageSize) {
            let group = DispatchGroup()
            
            group.enter()
            let offset = UIOffset(horizontal: CGFloat(settings.horizontalMargin), vertical: CGFloat(settings.verticalMargin))
            self.setRelativePageMargins(offset) {
                group.leave()
            }
            
            group.notify(queue: .main) {
                completionHandler?()
            }
        }
    }
    
    func setRelativePageMargins(_ pageMargins: UIOffset, completionHandler: (() -> Void)? = nil) {
       //relativePageMargins = pageMargins
        
        if let pageSize = self.pageSize {
            let horizontalMargin = floor(pageSize.width / 100 * pageMargins.horizontal)
            let verticalMargin = floor(pageSize.height / 100 * pageMargins.vertical)
            let absolutePageMargins = UIEdgeInsets(top: max(verticalMargin, webView.safeAreaInsets.top),
                                                   left: max(horizontalMargin, webView.safeAreaInsets.left),
                                                   bottom: max(verticalMargin, webView.safeAreaInsets.bottom),
                                                   right: max(horizontalMargin, webView.safeAreaInsets.right))
            setAbsolutePageMargins(absolutePageMargins, completionHandler: completionHandler)
        }
    }
    
    func setAbsolutePageMargins(_ pageMargins: UIEdgeInsets, completionHandler: (() -> Void)? = nil) {
        contentEdgeInsets = pageMargins
        
        let pageWidth = floor(pageSize!.width - pageMargins.left - pageMargins.right)
        let pageHeight = floor(pageSize!.height - pageMargins.top - pageMargins.bottom)
        let topInset = floor(pageMargins.top)
        let rightInset = floor(pageMargins.right)
        let bottomInset = floor(pageMargins.bottom)
        let leftInset = floor(pageMargins.left)

        let pageSizeScript = "setPageSettings(\(pageWidth),\(pageHeight),\(topInset),\(rightInset),\(bottomInset),\(leftInset))"
        
        guard self.pageSettingsString != pageSizeScript else {
            completionHandler?()
            return
        }
        
        webView.evaluateJavaScript(pageSizeScript) { (result, error) in
            completionHandler?()
        }
    }
    
}

